doc/adr/0016-end-to-end-testing-strategy.rst

```diff
@@ -0,0 +1,277 @@
+   Contract api, which would give a uniform interface for off-chain code such as different node
+   backends (private and public testnets, and emulator) and chain-indexer queries
+   (``cardano-node-client`` or Marconi in future).
+   It should also reduce the amount of boilerplate code and provide additional features such as
+   trace logging.
+
+.. _`argument 8`:
+
+8. There's value continuing to test ``cardano-cli`` with Plutus transactions for specific cli
+   flags and the cli's error handling with script evaluation.
+   Some examples of tests that should be covered:
+
+   * Cli flags that require use of Plutus scripts E.g. tx-out-reference-script-file or 
+     calculate-plutus-script-cost
+   * Cli behaviour when script evaluation passes. This could be displaying expected fee correctly.
+   * Cli behaviour when script evaluation fails. This can be how different types errors are formatted. 
+
+9. At some point, we may wish to incorporate the `cardano-node-emulator
+   <https://github.com/input-output-hk/Plutus-apps/pull/831>`_ as an alternative to ``cardano-testnet``.
+   This would enable us to run property based tests due to the node being much faster without consensus.
+   With CHaP, ``cardano-node-emulator`` would be released as a separate component, so no need to depend
+   on ``Plutus-apps``.
+
+10. We reserve the option of including additional packages to test from ``Plutus-apps`` at a later stage.
+
+Pros of building and maintaining our own test framework
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+* Plutus tools team will have full ownership of the end-to-end test environment and its priorities
+  for ``Plutus``.
+* Plutus scripts can be defined alongside the tests. ``cardano-node-tests`` requires
+  pre-compiled scripts.
+* Tests will demonstrate how these Haskell packages can be used together to guide Plutus application
+  development using the node apis. Particularly useful for less experienced Haskell developers.  
+* Possible to define tests once and run at different levels. For example, on private or public
+  testnets and with ``cardano-node-emulator`` emulated node.
+* Benefits from use of all ``Plutus`` apis. For example, using PlutusTx to produce scripts using a
+  typed interface, and optionally the Contract monad from ``Plutus-apps``.
+* Have the opportunity to add more components under test at a later stage, such as Marconi or a PAB.
+* ``cardano-cli`` would not be a dependency for Plutus test coverage so no risk of being blocked by its
+  stage of development.
+* Less dependence on repetitive manual approach for regression testing because tests can be planned
+  and implemented in parallel with feature implementation and integration.
+* Plutus team can implement and review majority of tests in Haskell rather than Python, which is
+  likely to be the team's preference. Also won't need to review as many tests in ``cardano-node-tests``.
```

Comment:
because there will be far less plutus-specific tests there.  If we decide not to proceed with this new Haskell framework then all plutus e2e testing requirements will need to be covered by `cardano-node-tests`.